modules/aws/sonar-upgrader/python_upgrader/upgrade/main.py

```diff
@@ -269,12 +268,44 @@ def run_test_connection_stage(args, extended_node_dict, upgrade_status_service):
             print(f"### Test connection to all DSF nodes succeeded")
 
 
-def run_collect_python_location_step(args, extended_node_dict, upgrade_status_service):
-    """
-    Collects the Python location in the DSF nodes and fills it in the extended_node_dict
-    """
-    if should_run_python(args):
-        collect_python_locations(extended_node_dict, args.stop_on_failure, upgrade_status_service)
+def run_collect_facts_step(args, extended_node_dict, upgrade_status_service):
+
+    for extended_node in extended_node_dict.values():
+        if upgrade_status_service.should_collect_facts(extended_node.get('dsf_node_id')):
+            upgrade_status_service.update_upgrade_status(extended_node.get('dsf_node_id'),
+                                                         UpgradeStatus.RUNNING_COLLECT_FACTS)
+            try:
+                collect_facts(extended_node)
+                upgrade_status_service.update_upgrade_status(extended_node.get('dsf_node_id'),
+                                                             UpgradeStatus.COLLECT_FACTS_SUCCEEDED)
+            except Exception as e:
+                upgrade_status_service.update_upgrade_status(extended_node.get('dsf_node_id'),
+                                                             UpgradeStatus.COLLECT_FACTS_FAILED)
+                if args.stop_on_failure:
+                    raise UpgradeException(
+                        f"Collecting Python location in {extended_node.get('dsf_node_name')} failed") from e
+                return None
```

Comment:
The is an issue in this return line, that if there is an error on the first node and stop_on_failue=false, then the code won't continue to the next node. I think it is better to keep the single and plural functions separated as it was before to avoid "return" bugs, for example: `collect_python_locations(**extended_node_dict**, stop_on_failure, upgrade_status_service)` and `collect_python_location(**extended_node**, stop_on_failure, upgrade_status_service)`

Reply:
I don't see how separating it would avoid issues like that. This is a bug in the multi function, and that doesn't seem related to if there was a single function separately.

Reply:
1. In case stop_on_failure=false and there is an error on one node, the code won't continue to the rest of the nodes like it should be due to the return statement. Please fix it and verify that the unit test pass (right now they are failing).
2. You are right it is not related to single and plural methods, it was a suggestion to avoid such issues in the future if someone will need to change this code

Reply:
Should be fixed, forgot to push